File: rail_fence_cipher.c
#include <stdio.h>
#include <string.h>

void encrypt(char *plainString, int key)
{

    char matrix[key][strlen(plainString)];
    int sign = 1;
    int j = 0;

    for (int i = 0; i < key; i++)
        for (int j = 0; j < strlen(plainString); j++)
            matrix[i][j] = '\0';

    for (int i = 0; i < strlen(plainString); i++)
    {
        matrix[j][i] = plainString[i];

        if (j == key - 1)
            sign = -1;
        else if (j == 0)
            sign = 1;
        j += sign;
    }

    int index = 0;
    for (int i = 0; i < key; i++)
        for (int j = 0; j < strlen(plainString); j++)
            if (matrix[i][j] != '\0')
                plainString[index++] = matrix[i][j];
}

int main()
{

    char plainString[] = "GeeksForGeeks";
    int key = 3;

    encrypt(plainString, key);

    printf("%s", plainString);

    return 0;
}

File: rail_fence_cipher.py
def encrypt(plain_text, key):

    cipher_text = ""

    matrix = [["" for _ in range(len(plain_text))] for _ in range(key)]

    j = 0
    for i in range(len(plain_text)):

        matrix[j][i] = plain_text[i]

        if j == 0:
            sign = 1
        if j == key - 1:
            sign = -1

        j += sign

    for row in range(key):
        for col in range(len(plain_text)):
            cipher_text += matrix[row][col]

    return cipher_text


def decrypt(cipher_text, key):

    plain_text = ""

    matrix = [["" for _ in range(len(cipher_text))] for _ in range(key)]

    j = 0
    for i in range(len(cipher_text)):

        matrix[j][i] = "*"

        if j == 0:
            sign = 1
        if j == key - 1:
            sign = -1

        j += sign

    k = 0
    for row in range(key):
        for col in range(len(cipher_text)):
            if matrix[row][col] == "*":
                matrix[row][col] = cipher_text[k]
                k += 1

    j = 0
    for i in range(len(cipher_text)):

        plain_text += matrix[j][i]

        if j == 0:
            sign = 1
        if j == key - 1:
            sign = -1

        j += sign

    return plain_text


plain_text = "GEEKSFORGEEKS"
key = 3

cipher_text = encrypt(plain_text, key)
print(cipher_text)

decrypted_cipher_text = decrypt(cipher_text, key)
print(decrypted_cipher_text)
